# ML Challenge 2026 — Business Entity Resolution

Match records from Source 2 and Source 3 to each deduplicated Source 1 entity.
Scored by **macro F_0.5** (precision weighted 2x recall), averaged over *every* S1
entity including singletons.

**Pipeline**

| Stage | What | Where |
|---|---|---|
| 0 | Normalize names/addresses -> Parquet | §2 |
| 1 | Blocking: 7 schemes -> candidate pairs | §4 |
| 2 | Pair features (39) | §5 |
| 3 | LightGBM pair scorer + calibration | §6 |
| 4 | Per-entity expected-F_0.5 set selection | §7 |
| 5 | Global exclusivity repair | §8 |
| — | Test inference + submission + validator | §10–11 |

**How to run.** Cells are top-to-bottom and each stage writes to `work/`, so you can
stop and resume. `CONFIG` in §1 controls whether you run on a fast 2% sample or the
full data — start with the sample (whole notebook ~4 min), then switch for the real
submission.

**Constraints:** no external data/APIs/geocoding (disqualification); final model must be
MIT/Apache-2.0 and <= 8B params. LightGBM is MIT; everything here is local.

## 1. Setup

In [2]:
import json, math, os, re, shutil, subprocess, sys, time, unicodedata
from dataclasses import dataclass, field
from pathlib import Path

import duckdb, numpy as np, pandas as pd, pyarrow as pa, pyarrow.parquet as pq
from rapidfuzz import fuzz
from rapidfuzz.distance import JaroWinkler

# ---------------------------------------------------------------- paths
# DATA_DIR is found by searching for the dataset, so it can live outside the repo.
NB_DIR = Path.cwd()
REPO_ROOT = next(p for p in [NB_DIR, *NB_DIR.parents] if (p / "code").is_dir())

def _find_data_dir():
    if os.environ.get("ER_DATA_DIR"):
        return Path(os.environ["ER_DATA_DIR"]).expanduser().resolve()
    for base in [REPO_ROOT, *list(REPO_ROOT.parents)[:3]]:
        for hit in base.rglob("test_source1.tsv"):
            return hit.parent.parent
    raise FileNotFoundError("set ER_DATA_DIR to the folder holding train/ and test/")

DATA_DIR = _find_data_dir()
WORK_DIR = Path(os.environ.get("ER_WORK_DIR", REPO_ROOT.parent / "work")).resolve()
OUTPUT_DIR = Path(os.environ.get("ER_OUTPUT_DIR", REPO_ROOT / "output")).resolve()
WORK_DIR.mkdir(parents=True, exist_ok=True)
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

SOURCES = [("train", 1, "train_source1.tsv"), ("train", 2, "train_source2.tsv"),
           ("train", 3, "train_source3.tsv"), ("test", 1, "test_source1.tsv"),
           ("test", 2, "test_source2.tsv"), ("test", 3, "test_source3.tsv")]
GROUND_TRUTH = "train_ground_truth.tsv"

# ---------------------------------------------------------------- knobs
CONFIG = dict(
    # Stage 0
    prepare_limit=None,      # rows per source; None = all 24.2M (~7 min)
    workers=max(1, os.cpu_count() - 2),

    # Stage 1 — per-scheme df ceilings. Measured: raising these to 5000 cost 8x the
    # compute for NO extra recall, and actually LOWERED recall@30 by crowding the cut.
    s1_sample=20,            # per-mille of S1 for the dev loop (20 = 2%)
    rare_per_s1=3,
    max_token_df=1000,
    skel_max_df=1000,
    addr_rare_per_s1=4, addr_max_df=300,
    pair_tokens=4, pair_per_s1=6, pair_max_df=2000,
    key_cap=300,             # cap for exact-key schemes
    scheme_cap=100,          # per-entity cap WITHIN a scheme, before the union
    top_k=30,                # final candidates per entity
    pair_budget=60_000_000,  # abort a scheme whose pre-flight estimate exceeds this

    # resources — this laptop has ~6GB free RAM and ~19GB free disk
    threads=6, mem_gb=5, spill_gb=12,
    test_shards=24,           # S1 shards for the full test run
)

print(f"repo   {REPO_ROOT}\ndata   {DATA_DIR}\nwork   {WORK_DIR}\noutput {OUTPUT_DIR}")
print(f"cpus {os.cpu_count()}  free disk {shutil.disk_usage(WORK_DIR).free / 2**30:.1f} GB")

repo   /home/subhrajit/Desktop/ML_Challenge_2026/ML_Challenge_2026
data   /home/subhrajit/Desktop/ML_Challenge_2026/6ab10eb3b23ba_student_resource/student_resource/dataset
work   /home/subhrajit/Desktop/ML_Challenge_2026/work
output /home/subhrajit/Desktop/ML_Challenge_2026/ML_Challenge_2026/output
cpus 12  free disk 14.3 GB


### 1.1 Confirm the raw files

Two quirks that bite naive loading, both real:

1. **`train_source1.tsv`'s header starts with a bare tab**, so its id column parses as
   `Unnamed: 0` while the other five files call it `entity_id`. `df['entity_id']` raises
   `KeyError` on that one file. We override the header everywhere.
2. Reading a `.tsv` without `sep='\t'` silently yields a single column.

In [3]:
SOURCE_COLUMNS = ["entity_id", "business_name", "business_address", "country"]
EXPECTED_ROWS = {"train_s1": 2_206_821, "train_s2": 5_034_616, "train_s3": 5_285_603,
                 "test_s1": 1_732_544, "test_s2": 4_887_273, "test_s3": 5_082_316}

for split, src, name in SOURCES:
    path = DATA_DIR / split / name
    with open(path, "rb") as fh:
        header = fh.readline().decode().rstrip("\n")
    flag = "  <-- unnamed first column!" if header.startswith("\t") else ""
    print(f"{split}/s{src}: {path.stat().st_size / 2**20:7.1f} MB  header={header!r}{flag}")

train/s1:   200.3 MB  header='\tbusiness_name\tbusiness_address\tcountry'  <-- unnamed first column!
train/s2:   466.6 MB  header='entity_id\tbusiness_name\tbusiness_address\tcountry'
train/s3:   480.4 MB  header='entity_id\tbusiness_name\tbusiness_address\tcountry'
test/s1:   166.9 MB  header='entity_id\tbusiness_name\tbusiness_address\tcountry'
test/s2:   485.9 MB  header='entity_id\tbusiness_name\tbusiness_address\tcountry'
test/s3:   482.6 MB  header='entity_id\tbusiness_name\tbusiness_address\tcountry'


## 2. EDA — what the data actually looks like

In [4]:
gt = pd.read_csv(DATA_DIR / "train" / GROUND_TRUTH, sep="\t",
                 dtype=str, keep_default_na=False, na_values=[])
empty = gt["matched_entity_ids"] == ""
counts = gt.loc[~empty, "matched_entity_ids"].str.count(",") + 1

print(f"S1 entities          {len(gt):,}")
print(f"singletons           {empty.sum():,} ({100 * empty.mean():.2f}%)")
print(f"mean matches/entity  {counts.sum() / len(gt):.2f}   max {counts.max()}")
print("\nmatches per entity:")
dist = counts.value_counts().sort_index()
print(f"  0: {empty.sum():>8,}")
for k, v in dist.items():
    print(f"  {k}: {v:>8,}")

S1 entities          2,206,821
singletons           123,247 (5.58%)
mean matches/entity  3.46   max 11

matches per entity:
  0:  123,247
  1:  119,157
  2:  375,212
  3:  530,841
  4:  484,115
  5:  321,957
  6:  164,868
  7:   63,968
  8:   18,680
  9:    4,205
  10:      534
  11:       37


### 2.1 The exclusivity property — verified, and it shapes Stage 5

If every S2/S3 record belongs to *at most one* S1 entity, then two entities claiming the
same id means at least one is wrong. Under a precision-heavy metric that is free score.
Worth checking rather than assuming:

In [5]:
mentions = gt.loc[~empty, "matched_entity_ids"].str.split(",").explode()
print(f"matched-id mentions {len(mentions):,}")
print(f"unique ids          {mentions.nunique():,}")
print("EXCLUSIVE: each S2/S3 record serves at most one S1 entity"
      if len(mentions) == mentions.nunique() else "NOT exclusive")

# Countries: France appears only in test, so nothing may be hardcoded to {US, India}.
for split, src, name in [("train", 1, "train_source1.tsv"), ("test", 1, "test_source1.tsv")]:
    c = pd.read_csv(DATA_DIR / split / name, sep="\t", header=0, names=SOURCE_COLUMNS,
                    usecols=["country"], dtype=str)["country"].value_counts()
    print(f"\n{split} S1 countries:\n{c.to_string()}")

matched-id mentions 7,638,365
unique ids          7,638,365
EXCLUSIVE: each S2/S3 record serves at most one S1 entity

train S1 countries:
country
US       1323633
India     883188

test S1 countries:
country
India     809986
US        663106
France    259452


## 3. Stage 0 — normalization

Entity resolution is mostly normalization quality; a good normalizer beats a fancy model.
This version fixes five defects found by measuring, not by reading:

- **Indic text was being deleted.** `NFKD` + `encode('ASCII','ignore')` folds `é`->`e`
  but Devanagari has no ASCII decomposition, so every character is dropped:
  `राम मार्केटिंग प्राइवेट लिमिटेड` -> `''`. Measured on 200k S3 rows, **2.016% of names
  became empty**; now 0.0010%. The data holds **nine** Indic scripts, not just Devanagari.
- **`\b&\b` never matches** (`&` is not a word char), so "Smith & Sons" and
  "Smith and Sons" stayed different. We drop the conjunction entirely.
- **Name and address were concatenated** into one vectorizer input, which lets a shared
  address carry an unrelated name — the worst error under F_0.5. Normalized separately.
- **Legal suffixes were dissolved** into the token soup, so `X Corp` vs `X LLC` looked
  like agreement. Kept as their own field.
- **Dotted acronyms shattered**: `S.A.R.L.` -> `l r s` (since `a` is a stopword).

In [6]:
# ----------------------------------------------------------- Indic scripts
_INDIC = re.compile(r"[ऀ-ൿ]")
_indic_run_re = re.compile(r"([ऀ-ൿ]+)")
_SCRIPT_BLOCKS = ((0x0900, 0x097F, "DEVANAGARI"), (0x0980, 0x09FF, "BENGALI"),
                  (0x0A00, 0x0A7F, "GURMUKHI"), (0x0A80, 0x0AFF, "GUJARATI"),
                  (0x0B00, 0x0B7F, "ORIYA"), (0x0B80, 0x0BFF, "TAMIL"),
                  (0x0C00, 0x0C7F, "TELUGU"), (0x0C80, 0x0CFF, "KANNADA"),
                  (0x0D00, 0x0D7F, "MALAYALAM"))

def _script_of(run):
    cp = ord(run[0])
    for lo, hi, name in _SCRIPT_BLOCKS:
        if lo <= cp <= hi:
            return name
    return "DEVANAGARI"

# IAST would give "praiveta"/"limiteda", sharing few n-grams with "private"/"limited".
# Mapping these first lands them on the same canonical token as the Latin sources.
_DEVA_PHRASE_MAP = {
    "प्राइवेट": "private", "प्रा": "private", "लिमिटेड": "limited", "लिमिडेट": "limited",
    "लि": "limited", "कंपनी": "company", "कम्पनी": "company", "एंड": "and", "एण्ड": "and",
    "इंडिया": "india", "भारत": "india", "सर्विसेज": "services", "सर्विस": "service",
    "इंटरप्राइजेज": "enterprises", "एंटरप्राइजेज": "enterprises", "उद्योग": "udyog",
    "ट्रेडर्स": "traders", "टेक्नोलॉजीज": "technologies", "सॉल्यूशंस": "solutions",
    "इंडस्ट्रीज": "industries", "मार्केटिंग": "marketing", "फाइनेंस": "finance",
    "स्टोर्स": "stores", "स्टोर": "store",
}
_deva_phrase_re = re.compile("|".join(sorted(map(re.escape, _DEVA_PHRASE_MAP),
                                             key=len, reverse=True)))

_IAST_FIXUPS = (("ṃ", "n"), ("ṁ", "n"),  # anusvara -> "infratech" not "imfratech"
                ("ph", "f"),                       # फ is written "f" far more often
                ("mg", "ng"))                      # "marketimg" -> "marketing"
# Tamil has no voiced/aspirated stops, so IAST renders them aspirated:
# "limidhedh" for "limited". Undoing that recovers most of the name.
_TAMIL_FIXUPS = (("dh", "t"), ("gh", "k"), ("bh", "p"))

# Applied ONLY to transliterated tokens, where output is too lossy for exact lookup.
# A blanket rule would also rewrite a genuine Latin name like "Limitless".
_TRANSLIT_SUFFIX_RULES = (
    (re.compile(r"^limi"), "limited"),
    (re.compile(r"^(?:praiv|piraiv|privat|praiw)"), "private"),
    (re.compile(r"^(?:ailaail|elel|elail)"), "llp"),
    (re.compile(r"^(?:korpor|corpor)"), "corporation"),
    (re.compile(r"^(?:kampani|kampeni|kampan)"), "company"),
)

def fold_ascii(text):
    """Strip diacritics. Safe only AFTER transliteration."""
    return unicodedata.normalize("NFKD", text).encode("ASCII", "ignore").decode("ascii")

def _fix_indic_word(word, script):
    # Fold first: IAST emits "ḍh"/"ā", so the patterns below would never match.
    word = fold_ascii(word).lower()
    if script == "TAMIL":
        for a, b in _TAMIL_FIXUPS:
            word = word.replace(a, b)
    for a, b in _IAST_FIXUPS:
        word = word.replace(a, b)
    # Inherent final vowel: "rama" -> "ram". Devanagari only — applying it to the
    # others ate real vowels ("alpha" -> "alf").
    if script == "DEVANAGARI" and len(word) > 3 and word.endswith("a"):
        word = word[:-1]
    for pattern, canon in _TRANSLIT_SUFFIX_RULES:
        if pattern.match(word):
            return canon
    return word

def _transliterate_indic(text):
    """Indic -> Latin, run by run, so Latin words already present are untouched."""
    text = _deva_phrase_re.sub(lambda m: " " + _DEVA_PHRASE_MAP[m.group(0)] + " ", text)
    if not _INDIC.search(text):
        return text
    try:
        from indic_transliteration import sanscript
        from indic_transliteration.sanscript import transliterate
    except Exception:
        return text  # never let a transliteration failure blank a record
    out = []
    for i, part in enumerate(_indic_run_re.split(text)):
        if i % 2 == 0:
            out.append(part); continue
        script = _script_of(part)
        try:
            latin = transliterate(part, getattr(sanscript, script), sanscript.IAST)
        except Exception:
            out.append(part); continue
        out.append(" ".join(_fix_indic_word(w, script) for w in latin.split()))
    return "".join(out)

In [7]:
# ----------------------------------------------------------- vocabularies
LEGAL_SUFFIX_CANON = {
    "corporation": "corp", "corp": "corp", "corpn": "corp",
    "incorporated": "inc", "inc": "inc",
    "private": "pvt", "pvt": "pvt", "pte": "pvt",
    "limited": "ltd", "ltd": "ltd", "ltda": "ltd",
    "llp": "llp", "llc": "llc", "lp": "lp",
    "company": "co", "co": "co", "cos": "co",
    "sarl": "sarl", "sas": "sas", "sasu": "sas", "sa": "sa", "eurl": "eurl", "snc": "snc",
    "gmbh": "gmbh", "plc": "plc", "ag": "ag", "bv": "bv", "nv": "nv",
    "partnership": "partnership", "associates": "associates", "assoc": "associates",
    "holdings": "holdings", "group": "group", "trust": "trust", "society": "society",
}
NOISE_TOKENS = {"and", "the", "of", "a", "an"}
ADDR_ABBREV = {
    "rd": "road", "road": "road", "st": "street", "str": "street", "street": "street",
    "ave": "avenue", "av": "avenue", "avenue": "avenue", "blvd": "boulevard",
    "boulevard": "boulevard", "bd": "boulevard", "ln": "lane", "lane": "lane",
    "dr": "drive", "drive": "drive", "ct": "court", "court": "court", "pl": "place",
    "place": "place", "sq": "square", "square": "square", "hwy": "highway",
    "highway": "highway", "pkwy": "parkway", "parkway": "parkway", "apt": "apartment",
    "apartment": "apartment", "ste": "suite", "suite": "suite", "bldg": "building",
    "building": "building", "flr": "floor", "fl": "floor", "floor": "floor",
    "nr": "near", "near": "near", "opp": "opposite", "opposite": "opposite",
    "n": "north", "north": "north", "s": "south", "south": "south", "e": "east",
    "east": "east", "w": "west", "west": "west", "ne": "northeast", "nw": "northwest",
    "se": "southeast", "sw": "southwest", "marg": "marg", "nagar": "nagar",
    "colony": "colony", "sector": "sector", "phase": "phase", "block": "block",
    "gali": "gali", "chowk": "chowk", "po": "postoffice", "ps": "policestation",
    "dist": "district", "tq": "taluk", "taluk": "taluk", "tehsil": "taluk",
    "r": "rue", "rue": "rue",
}
ADDR_NOISE = {"apartment", "suite", "floor", "building", "unit", "no", "number",
              "near", "opposite", "postoffice", "policestation"}

_word_re = re.compile(r"[a-z0-9]+")
_amp_re = re.compile(r"&amp;|&#38;|&")
_zerowidth_re = re.compile(r"[​-‍]")
_acronym_re = re.compile(r"\b[a-z](?:\.[a-z])+\.?")   # "s.a.r.l." -> one token
_pin6_re, _zip5_re = re.compile(r"\b\d{6}\b"), re.compile(r"\b\d{5}\b")

def _pre_clean(raw):
    if raw is None:
        return ""
    text = str(raw)
    if not text or text == "nan":
        return ""
    text = unicodedata.normalize("NFKC", text)
    text = _zerowidth_re.sub("", text)       # ZWNJ/ZWJ appear inside Indic names
    if _INDIC.search(text):
        text = _transliterate_indic(text)
    text = _amp_re.sub(" and ", text)
    text = fold_ascii(text).lower()
    return _acronym_re.sub(lambda m: m.group(0).replace(".", ""), text)

@dataclass(slots=True)
class NameParts:
    core: list = field(default_factory=list)     # discriminative tokens
    suffix: list = field(default_factory=list)   # canonical legal suffixes
    key: str = ""                                # order-invariant block key
    @property
    def text(self): return " ".join(self.core)

def normalize_name(raw):
    text = _pre_clean(raw)
    if not text:
        return NameParts()
    core, suffix = [], []
    for tok in _word_re.findall(text):
        canon = LEGAL_SUFFIX_CANON.get(tok)
        if canon is not None:
            if canon not in suffix:
                suffix.append(canon)
            continue
        if tok in NOISE_TOKENS:
            continue
        core.append(tok)
    if not core and suffix:          # a name that is ONLY a suffix would block to nothing
        core = list(suffix)
    return NameParts(core, suffix, " ".join(sorted(set(core))))

@dataclass(slots=True)
class AddressParts:
    tokens: list = field(default_factory=list)
    numbers: list = field(default_factory=list)
    postcode: str = ""
    is_empty: bool = True
    @property
    def text(self): return " ".join(self.tokens)

def normalize_address(raw):
    text = _pre_clean(raw)
    if not text.strip():
        return AddressParts()
    m = _pin6_re.search(text) or _zip5_re.search(text)   # before tokenizing
    postcode = m.group(0) if m else ""
    tokens, numbers = [], []
    for tok in _word_re.findall(text):
        if tok == postcode:
            continue
        if tok.isdigit() or (any(c.isdigit() for c in tok) and any(c.isalpha() for c in tok)):
            if tok not in numbers:
                numbers.append(tok)
            continue
        tok = ADDR_ABBREV.get(tok, tok)
        if tok in NOISE_TOKENS or tok in ADDR_NOISE:
            continue
        tokens.append(tok)
    return AddressParts(tokens, numbers, postcode, False)

### 3.1 Self-test — the Stage 0 gate

Transliterated pairs are graded on a **similarity threshold**, not key equality:
transliteration is lossy (`സിൽവർ കൺസൾട്ടൻസി` -> `siva kasattasi`), so a test demanding
exact keys there would be a test that lies. Negative cases guard against
over-normalizing two genuinely different businesses onto one key.

In [8]:
FUZZY_MIN = 60
NAME_CASES = [
    ("Ram Marketing Private Limited", "राम मार्केटिंग प्राइवेट लिमिटेड", "key"),
    ("मॉडर्न फाइनेंस", "Modern Finance", "fuzzy"),
    ("श्री साई इंफ्राटेक", "Shri Sai Infratech", "fuzzy"),
    ("Smith & Sons Corp", "Smith and Sons Corporation", "key"),
    ("Smith & Sons", "Smith Sons", "key"),
    ("Café Déjà Inc", "Cafe Deja Incorporated", "key"),
    ("Acme Traders Pvt Ltd", "Traders Acme Private Limited", "key"),
    ("Marina Ecole France Sarl", "Marina Ecole France S.A.R.L.", "key"),
    ("Zephay Labs Inc", "Zephyr Labs Inc", "differ"),
    ("Prime Money", "Prime Motors", "differ"),
]
INDIC_SAMPLES = [
    ("Devanagari", "राम मार्केटिंग प्राइवेट लिमिटेड"), ("Devanagari", "मॉडर्न फाइनेंस"),
    ("Malayalam", "സിൽവർ കൺസൾട്ടൻസി പ്രൈവറ്റ് ലിമിറ്റഡ്"),
    ("Gujarati", "આલ્ફા ફાઉન્ડેશન પ્રાઇવેટ લિમિટેડ"),
    ("Kannada", "ರಾಮ್ ಬಿಸಿನೆಸ್ ಪ್ರೈವೇಟ್ ಲಿಮಿಟೆಡ್"),
    ("Tamil", "ஸ்மார்ட் எக்ஸ்போர்ட்ஸ் பிரைவேட் லிமிடெட்"),
    ("Bengali", "রেড টেক প্রাইভেট লিমিটেড"),
    ("Telugu", "లోటస్ మార్కెటింగ్ ప్రైవేట్ లిమిటెడ్"),
    ("Gurmukhi", "ਸ਼ਿਵਮ ਗੋਲਡਨ ਪ੍ਰੋਡਿਊਸਰ ਐਲਐਲਪੀ"), ("Odia", "ସୁପର୍ ମ୍ୟାନେଜମେଣ୍ଟ୍"),
]

fails = 0
print("=== name normalization ===")
for a, b, mode in NAME_CASES:
    na, nb = normalize_name(a), normalize_name(b)
    same = na.key == nb.key and na.key != ""
    sim = fuzz.token_set_ratio(na.text, nb.text)
    ok = same if mode == "key" else (sim >= FUZZY_MIN if mode == "fuzzy" else not same)
    fails += not ok
    note = "key match" if mode == "key" else (f"sim={sim:.0f}" if mode == "fuzzy" else "keys differ")
    print(f"  [{'ok ' if ok else 'FAIL'}] {mode:6s} {note:12s} {na.key!r} | {nb.key!r}")

print("\n=== non-empty guarantee, all 9 Indic scripts ===")
for script, raw in INDIC_SAMPLES:
    p = normalize_name(raw)
    fails += not p.core
    print(f"  [{'ok ' if p.core else 'FAIL'}] {script:11s} core={p.core} suffix={p.suffix}")

print("\n=== address slots ===")
for raw in ["1795 Westchester Drive, High Point, NC",
            "No 10 Enkay Square, 448A, Udyog Vihar Phase V, Gurugram, HR",
            "63 R. DE DIEPPE, LILLE, Hauts-de-France",
            "Near SBI ATM, Opp. Bus Stand, Nashik 422001", ""]:
    p = normalize_address(raw)
    print(f"  {raw[:46]!r:48s} tok={p.tokens[:4]} nums={p.numbers} pin={p.postcode!r}")

print(f"\n{'PASS' if fails == 0 else f'{fails} FAILURE(S)'}")

=== name normalization ===
  [ok ] key    key match    'marketing ram' | 'marketing ram'
  [ok ] fuzzy  sim=86       'finance madarn' | 'finance modern'
  [ok ] fuzzy  sim=82       'imfratek sai sri' | 'infratech sai shri'
  [ok ] key    key match    'smith sons' | 'smith sons'
  [ok ] key    key match    'smith sons' | 'smith sons'
  [ok ] key    key match    'cafe deja' | 'cafe deja'
  [ok ] key    key match    'acme traders' | 'acme traders'
  [ok ] key    key match    'ecole france marina' | 'ecole france marina'
  [ok ] differ keys differ  'labs zephay' | 'labs zephyr'
  [ok ] differ keys differ  'money prime' | 'motors prime'

=== non-empty guarantee, all 9 Indic scripts ===
  [ok ] Devanagari  core=['ram', 'marketing'] suffix=['pvt', 'ltd']
  [ok ] Devanagari  core=['madarn', 'finance'] suffix=[]
  [ok ] Malayalam   core=['siva', 'kasattasi'] suffix=['pvt', 'ltd']
  [ok ] Gujarati    core=['alfa', 'faundesana'] suffix=['pvt', 'ltd']
  [ok ] Kannada     core=['ram', 'bisines'] su

### 3.2 Run Stage 0 over the sources

Streamed in chunks and fanned across cores — never load a whole source into pandas
(2.5 GB of TSV becomes 8–12 GB of Python strings, against ~6 GB free RAM).
Set `CONFIG['prepare_limit']=200_000` for a fast pass; `None` does all 24.2M rows (~7 min).

Skips any source already written, so re-running this cell is cheap.

In [9]:
import multiprocessing as mp

SCHEMA = pa.schema([
    ("entity_id", pa.string()), ("country", pa.string()), ("source", pa.int8()),
    ("name_key", pa.string()), ("name_text", pa.string()),
    ("name_tokens", pa.list_(pa.string())), ("name_suffix", pa.string()),
    ("n_core", pa.int16()), ("addr_text", pa.string()),
    ("addr_tokens", pa.list_(pa.string())), ("addr_nums", pa.list_(pa.string())),
    ("postcode", pa.string()), ("addr_empty", pa.bool_()),
])

def _normalize_batch(batch):
    out = []
    for raw_name, raw_addr in batch:
        n, a = normalize_name(raw_name), normalize_address(raw_addr)
        out.append((n.key, n.text, n.core, " ".join(n.suffix), len(n.core),
                    a.text, a.tokens, a.numbers, a.postcode, a.is_empty))
    return out

def prepare_source(tsv, source, out, pool, workers, limit=None):
    writer = pq.ParquetWriter(out, SCHEMA, compression="zstd")
    total, t0 = 0, time.perf_counter()
    try:
        # header=0 + names= : train_source1.tsv's id column is unnamed (leading tab)
        for chunk in pd.read_csv(tsv, sep="\t", header=0, names=SOURCE_COLUMNS,
                                 chunksize=200_000, dtype=str, keep_default_na=False,
                                 na_values=[], nrows=limit):
            rows = list(zip(chunk["business_name"], chunk["business_address"]))
            size = max(1, len(rows) // workers + 1)
            parts = pool.map(_normalize_batch,
                             [rows[i:i + size] for i in range(0, len(rows), size)])
            norm = [r for p in parts for r in p]
            writer.write_table(pa.table({
                "entity_id": pa.array(chunk["entity_id"].tolist(), pa.string()),
                "country": pa.array(chunk["country"].tolist(), pa.string()),
                "source": pa.array([source] * len(chunk), pa.int8()),
                "name_key": pa.array([r[0] for r in norm], pa.string()),
                "name_text": pa.array([r[1] for r in norm], pa.string()),
                "name_tokens": pa.array([r[2] for r in norm], pa.list_(pa.string())),
                "name_suffix": pa.array([r[3] for r in norm], pa.string()),
                "n_core": pa.array([r[4] for r in norm], pa.int16()),
                "addr_text": pa.array([r[5] for r in norm], pa.string()),
                "addr_tokens": pa.array([r[6] for r in norm], pa.list_(pa.string())),
                "addr_nums": pa.array([r[7] for r in norm], pa.list_(pa.string())),
                "postcode": pa.array([r[8] for r in norm], pa.string()),
                "addr_empty": pa.array([r[9] for r in norm], pa.bool_()),
            }, schema=SCHEMA))
            total += len(chunk)
    finally:
        writer.close()
    print(f"  {split}/s{source}: {total:>10,} rows in {time.perf_counter() - t0:5.1f}s")
    return total

with mp.Pool(CONFIG["workers"]) as pool:
    for split, source, name in SOURCES:
        out = WORK_DIR / f"{split}_s{source}.parquet"
        if out.exists() and CONFIG["prepare_limit"] is None:
            print(f"  {split}/s{source}: exists ({pq.ParquetFile(out).metadata.num_rows:,} rows), skip")
            continue
        prepare_source(DATA_DIR / split / name, source, out, pool,
                       CONFIG["workers"], CONFIG["prepare_limit"])

print("\nrow-count check:")
for key, expected in EXPECTED_ROWS.items():
    n = pq.ParquetFile(WORK_DIR / f"{key}.parquet").metadata.num_rows
    print(f"  {key:9s} {n:>10,}  expected {expected:>10,}  {'OK' if n == expected else 'MISMATCH'}")

  train/s1: exists (2,206,821 rows), skip
  train/s2: exists (5,034,616 rows), skip
  train/s3: exists (5,285,603 rows), skip
  test/s1: exists (1,732,544 rows), skip
  test/s2: exists (4,887,273 rows), skip
  test/s3: exists (5,082,316 rows), skip

row-count check:
  train_s1   2,206,821  expected  2,206,821  OK
  train_s2   5,034,616  expected  5,034,616  OK
  train_s3   5,285,603  expected  5,285,603  OK
  test_s1    1,732,544  expected  1,732,544  OK
  test_s2    4,887,273  expected  4,887,273  OK
  test_s3    5,082,316  expected  5,082,316  OK


## 4. The metric — derive it before optimizing it

With `k` predicted, `m` true, `h` correct:

$$P = h/k,\quad R = h/m \quad\Rightarrow\quad F_{0.5} = \frac{1.25\,PR}{0.25P + R} = \frac{1.25\,h}{k + 0.25\,m}$$

That reduced form is the useful one: the score is **linear in hits**, and each wrong
guess costs exactly one unit of `k`. Edge cases follow the rules: empty-vs-empty = 1.0
(a correct singleton), and any prediction on a true singleton = 0.0.

In [10]:
def f05(pred: set, true: set) -> float:
    k, m = len(pred), len(true)
    if k == 0 and m == 0:
        return 1.0
    if k == 0 or m == 0:
        return 0.0
    h = len(pred & true)
    return 0.0 if h == 0 else 1.25 * h / (k + 0.25 * m)

def macro_f05(pred: dict, truth: dict) -> float:
    if not truth:
        return 0.0
    return sum(f05(pred.get(s, set()), t) for s, t in truth.items()) / len(truth)

def load_ground_truth(s1_ids=None) -> dict:
    g = pd.read_csv(DATA_DIR / "train" / GROUND_TRUTH, sep="\t",
                    dtype=str, keep_default_na=False, na_values=[])
    out = {}
    for sid, matched in zip(g["source1_entity_id"], g["matched_entity_ids"]):
        if s1_ids is not None and sid not in s1_ids:
            continue
        out[sid] = set(matched.split(",")) if matched else set()
    return out

# The README's worked example, which is how we know the spec was read correctly.
ex = f05({"S2-00047", "S2-00193", "S3-00812"}, {"S2-00047", "S3-00812"})
assert abs(ex - 0.714) < 0.001, ex
assert f05(set(), set()) == 1.0 and f05({"x"}, set()) == 0.0 and f05({"a"}, {"a"}) == 1.0
print(f"metric checks PASS (README example = {ex:.3f})")

metric checks PASS (README example = 0.714)


## 5. Stage 1 — blocking

Brute force is `1,732,544 x 9,969,589 ~ 1.7e13` pairs, so blocking is not an
optimization, it is the only way this is solvable. **Blocking recall is a hard ceiling
on the final score.**

Seven schemes, unioned. Every join is `within country` — written generically over the
column, never hardcoded to `{US, India}`, so the 259,452 French test entities work.

| # | Key | Catches | Why it exists |
|---|---|---|---|
| A | exact `name_key` | clean duplicates | cheapest, highest precision |
| A2 | concatenated name, domains stripped | `cultural association` = `culturalassociation` | sources that collapse whitespace |
| B | shared rare token (IDF-scored) | word order, partial names | the original workhorse |
| C | shared rare consonant skeleton | transliteration drift | `consulting`/`kansalting` -> `knsltng` |
| D | postcode + shared name token | address-led | weak; kept for provenance |
| E | shared rare **address** tokens, no name needed | names that share *nothing* | `bauer international`/`novimira` |
| F | conjunctive **token pairs** | all-common-token names | `digital products`/`sri digital products` |

Two cost rules, both learned by breaking things:

1. **A separate `df` ceiling per scheme.** Skeletons are denser than tokens *by design*,
   so reusing the token ceiling spilled 15 GB and took the disk to 96%.
2. **Cap each scheme per entity BEFORE the union.** Ranking the full ~100M-row union in
   one window exhausted memory and the spill budget together.

In [11]:
DOMAIN_TOKENS = "('com','www','net','org','co','in','io','biz','info','html')"
SQUASH_SQL = f"array_to_string(list_filter(name_tokens, t -> t NOT IN {DOMAIN_TOKENS}), '')"
# sh->s, ph->f, c->k, then drop vowels: transliteration mangles vowels but largely
# preserves consonants, so this turns fuzzy similarity into an exact equi-join.
SKELETON_SQL = ("regexp_replace(regexp_replace(regexp_replace(regexp_replace("
                "tok,'sh','s','g'),'ph','f','g'),'c','k','g'),'[aeiou]','','g')")

def connect(threads=None, mem_gb=None, spill_gb=None, db=None):
    """Connect with a HARD spill ceiling so a bad config fails instead of filling the disk.

    `db` gives a FILE-backed database. The pool index (cand_pair alone is ~40M rows)
    stays resident for every shard query, which does not fit in the memory budget;
    on disk it is paged in as needed instead of competing with the shard's own joins.
    """
    threads = threads or CONFIG["threads"]; mem_gb = mem_gb or CONFIG["mem_gb"]
    spill_gb = spill_gb or CONFIG["spill_gb"]
    con = duckdb.connect(str(db) if db else ":memory:")
    tmp = WORK_DIR / "duckdb_tmp"; tmp.mkdir(parents=True, exist_ok=True)
    free_gb = shutil.disk_usage(tmp).free / 2**30
    if free_gb < spill_gb + 2:
        spill_gb = max(1, int(free_gb) - 2)
        print(f"  ! only {free_gb:.1f}GB free, capping spill at {spill_gb}GB")
    con.execute(f"SET threads={threads}")
    con.execute(f"SET memory_limit='{mem_gb}GB'")
    con.execute(f"SET temp_directory='{tmp}'")
    con.execute(f"SET max_temp_directory_size='{spill_gb}GB'")
    con.execute("SET preserve_insertion_order=false")
    return con

def estimate_pairs(con, table, label, budget=None):
    """Exact pre-flight pair count: each (entity,key) row fans out to `df` rows,
    so SUM(df) IS the join's output size. Costs milliseconds."""
    budget = budget or CONFIG["pair_budget"]
    est = int(con.sql(f"SELECT COALESCE(SUM(df),0) FROM {table}").fetchone()[0] or 0)
    print(f"  estimate {label:<10s} {est:>14,} pairs  [{'OK' if est <= budget else 'OVER BUDGET'}]")
    if est > budget:
        raise RuntimeError(f"{label}: ~{est:,} pairs exceeds budget {budget:,}. "
                           f"Lower its df ceiling or rare-per-s1.")
    return est

def _step(con, label, sql):
    t = time.perf_counter(); con.execute(sql)
    print(f"  {label:<22s} {time.perf_counter() - t:6.1f}s", flush=True)

### 5.1 Pool-side index — built once

These tables depend only on the S2/S3 pool, not on which S1 rows we are matching, so
they are built once and reused by every shard. Rebuilding them per shard wasted ~90s
each time and left no memory for the shard work.

In [12]:
def build_pool(con, split):
    pool = [str(WORK_DIR / f"{split}_s2.parquet"), str(WORK_DIR / f"{split}_s3.parquet")]
    con.execute(f"""CREATE OR REPLACE VIEW cand AS
        SELECT entity_id, country, name_key, name_tokens, addr_tokens, postcode
        FROM read_parquet({pool})""")
    n_cand = con.sql("SELECT COUNT(*) FROM cand").fetchone()[0]
    print(f"  pool {n_cand:,} records")

    _step(con, "name tokens", """
        CREATE OR REPLACE TABLE cand_tok AS
        SELECT entity_id, country, UNNEST(name_tokens) AS tok FROM cand""")
    _step(con, "token df", f"""
        CREATE OR REPLACE TABLE tokdf AS
        SELECT country, tok, COUNT(*)::BIGINT AS df,
               LN(1.0 + {n_cand}.0 / COUNT(*)) AS idf
        FROM cand_tok GROUP BY country, tok""")
    _step(con, "squash keys", f"""
        CREATE OR REPLACE TABLE cand_keys AS
        SELECT entity_id, country, {SQUASH_SQL} AS squash FROM cand""")
    _step(con, "skeletons", f"""
        CREATE OR REPLACE TABLE cand_skel AS
        SELECT entity_id, country, skel FROM (
            SELECT entity_id, country, {SKELETON_SQL} AS skel FROM cand_tok
        ) WHERE length(skel) >= 3;
        CREATE OR REPLACE TABLE skeldf AS
        SELECT country, skel, COUNT(*)::BIGINT AS df,
               LN(1.0 + {n_cand}.0 / COUNT(*)) AS idf
        FROM cand_skel GROUP BY country, skel;""")
    _step(con, "address tokens", f"""
        CREATE OR REPLACE TABLE cand_atok AS
        SELECT entity_id, country, UNNEST(addr_tokens) AS atok FROM cand;
        CREATE OR REPLACE TABLE atokdf AS
        SELECT country, atok, COUNT(*)::BIGINT AS df,
               LN(1.0 + {n_cand}.0 / COUNT(*)) AS idf
        FROM cand_atok GROUP BY country, atok;""")
    _step(con, "token pairs", f"""
        CREATE OR REPLACE TABLE cand_rare AS
        SELECT entity_id, country, tok FROM (
            SELECT c.entity_id, c.country, c.tok,
                   ROW_NUMBER() OVER (PARTITION BY c.entity_id ORDER BY t.df ASC, c.tok) AS rk
            FROM cand_tok c JOIN tokdf t ON t.country = c.country AND t.tok = c.tok
        ) WHERE rk <= {CONFIG['pair_tokens']};
        CREATE OR REPLACE TABLE cand_pair AS
        SELECT a.entity_id, a.country, a.tok || '|' || b.tok AS pk
        FROM cand_rare a JOIN cand_rare b
          ON b.entity_id = a.entity_id AND b.tok > a.tok;
        CREATE OR REPLACE TABLE pairdf AS
        SELECT country, pk, COUNT(*)::BIGINT AS df,
               LN(1.0 + {n_cand}.0 / COUNT(*)) AS idf
        FROM cand_pair GROUP BY country, pk;""")
    return n_cand

### 5.2 Shard-side — the seven schemes

`s1_filter` picks which S1 rows this pass handles: a per-mille sample for the dev loop,
or one shard of the full split. **The pool is never sampled**, so measured recall is not
inflated by an easy subset.

In [13]:
def build_shard(con, split, s1_filter, suffix="", top_k=None):
    cfg, top_k = CONFIG, top_k or CONFIG["top_k"]
    s1p = str(WORK_DIR / f"{split}_s1.parquet")
    con.execute(f"""CREATE OR REPLACE VIEW s1 AS
        SELECT entity_id, country, name_key, name_tokens, addr_tokens, postcode
        FROM read_parquet('{s1p}') {s1_filter}""")
    n_s1 = con.sql("SELECT COUNT(*) FROM s1").fetchone()[0]
    print(f"  S1 rows {n_s1:,}")

    # ---- key tables, each keeping only this entity's RAREST few keys ----
    _step(con, "s1 rare tokens", f"""
        CREATE OR REPLACE TABLE s1_rare AS
        SELECT entity_id, country, tok, df, idf FROM (
            SELECT s.entity_id, s.country, t.tok, t.df, t.idf,
                   ROW_NUMBER() OVER (PARTITION BY s.entity_id ORDER BY t.df ASC, t.tok) AS rk
            FROM (SELECT entity_id, country, UNNEST(name_tokens) AS tok FROM s1) s
            JOIN tokdf t ON t.country = s.country AND t.tok = s.tok
        ) WHERE rk <= {cfg['rare_per_s1']} AND df <= {cfg['max_token_df']}""")
    estimate_pairs(con, "s1_rare", "scheme B")

    _step(con, "s1 skeletons", f"""
        CREATE OR REPLACE TABLE s1_skel AS
        SELECT entity_id, country, skel, df, idf FROM (
            SELECT s.entity_id, s.country, k.skel, k.df, k.idf,
                   ROW_NUMBER() OVER (PARTITION BY s.entity_id ORDER BY k.df ASC, k.skel) AS rk
            FROM (SELECT entity_id, country, {SKELETON_SQL} AS skel
                  FROM (SELECT entity_id, country, UNNEST(name_tokens) AS tok FROM s1)) s
            JOIN skeldf k ON k.country = s.country AND k.skel = s.skel
            WHERE length(s.skel) >= 3
        ) WHERE rk <= {cfg['rare_per_s1']} AND df <= {cfg['skel_max_df']}""")
    estimate_pairs(con, "s1_skel", "scheme C")

    _step(con, "s1 rare addr", f"""
        CREATE OR REPLACE TABLE s1_arare AS
        SELECT entity_id, country, atok, df, idf FROM (
            SELECT s.entity_id, s.country, t.atok, t.df, t.idf,
                   ROW_NUMBER() OVER (PARTITION BY s.entity_id ORDER BY t.df ASC, t.atok) AS rk
            FROM (SELECT entity_id, country, UNNEST(addr_tokens) AS atok FROM s1) s
            JOIN atokdf t ON t.country = s.country AND t.atok = s.atok
        ) WHERE rk <= {cfg['addr_rare_per_s1']} AND df <= {cfg['addr_max_df']}""")
    estimate_pairs(con, "s1_arare", "scheme E")

    _step(con, "s1 token pairs", f"""
        CREATE OR REPLACE TABLE s1_rare4 AS
        SELECT entity_id, country, tok FROM (
            SELECT s.entity_id, s.country, s.tok,
                   ROW_NUMBER() OVER (PARTITION BY s.entity_id ORDER BY t.df ASC, s.tok) AS rk
            FROM (SELECT entity_id, country, UNNEST(name_tokens) AS tok FROM s1) s
            JOIN tokdf t ON t.country = s.country AND t.tok = s.tok
        ) WHERE rk <= {cfg['pair_tokens']};
        CREATE OR REPLACE TABLE s1_pair AS
        SELECT entity_id, country, pk, df, idf FROM (
            SELECT sp.entity_id, sp.country, sp.pk, d.df, d.idf,
                   ROW_NUMBER() OVER (PARTITION BY sp.entity_id ORDER BY d.df ASC, sp.pk) AS rk
            FROM (SELECT a.entity_id, a.country, a.tok || '|' || b.tok AS pk
                  FROM s1_rare4 a JOIN s1_rare4 b
                    ON b.entity_id = a.entity_id AND b.tok > a.tok) sp
            JOIN pairdf d ON d.country = sp.country AND d.pk = sp.pk
        ) WHERE rk <= {cfg['pair_per_s1']} AND df <= {cfg['pair_max_df']}""")
    estimate_pairs(con, "s1_pair", "scheme F")

    # ---- the schemes ----
    # A's cap used to keep an ARBITRARY subset (ORDER BY cand_id), losing true matches
    # for generic names: identical pairs like 'united network'/'united network'.
    # Postcode agreement is a meaningful tiebreak.
    _step(con, "A exact key", f"""
        CREATE OR REPLACE TABLE pairs_a AS
        SELECT s1_id, cand_id FROM (
            SELECT s.entity_id AS s1_id, c.entity_id AS cand_id,
                   ROW_NUMBER() OVER (PARTITION BY s.entity_id ORDER BY
                       (c.postcode <> '' AND c.postcode = s.postcode) DESC, c.entity_id) AS rk
            FROM s1 s JOIN cand c ON c.country = s.country AND c.name_key = s.name_key
            WHERE s.name_key <> ''
        ) WHERE rk <= {cfg['key_cap']}""")
    _step(con, "A2 squash", f"""
        CREATE OR REPLACE TABLE pairs_a2 AS
        SELECT s1_id, cand_id FROM (
            SELECT s.entity_id AS s1_id, c.entity_id AS cand_id,
                   ROW_NUMBER() OVER (PARTITION BY s.entity_id ORDER BY c.entity_id) AS rk
            FROM (SELECT entity_id, country, {SQUASH_SQL} AS squash FROM s1) s
            JOIN cand_keys c ON c.country = s.country AND c.squash = s.squash
            WHERE length(s.squash) >= 4
        ) WHERE rk <= {cfg['key_cap']}""")

    # Each scheme capped per entity BEFORE the union. scheme_cap > top_k on purpose,
    # so the union still has candidates to choose between.
    def ranked(name, key_tbl, key_col, cand_tbl, having=""):
        _step(con, name, f"""
            CREATE OR REPLACE TABLE pairs_{name.split()[0].lower()} AS
            SELECT s1_id, cand_id, idf_score, shared FROM (
                SELECT s1_id, cand_id, idf_score, shared,
                       ROW_NUMBER() OVER (PARTITION BY s1_id
                           ORDER BY idf_score DESC, shared DESC, cand_id) AS rk
                FROM (
                    SELECT r.entity_id AS s1_id, c.entity_id AS cand_id,
                           SUM(r.idf) AS idf_score, COUNT(*)::INT AS shared
                    FROM {key_tbl} r JOIN {cand_tbl} c
                      ON c.country = r.country AND c.{key_col} = r.{key_col}
                    GROUP BY 1, 2 {having}
                )
            ) WHERE rk <= {cfg['scheme_cap']}""")

    ranked("b rare token", "s1_rare", "tok", "cand_tok")
    ranked("c skeleton", "s1_skel", "skel", "cand_skel")
    # >=2 shared rare address tokens: one address token alone is weak evidence.
    ranked("e rare addr", "s1_arare", "atok", "cand_atok", "HAVING COUNT(*) >= 2")
    ranked("f token pair", "s1_pair", "pk", "cand_pair")

    _step(con, "D postcode", """
        CREATE OR REPLACE TABLE pairs_d AS
        SELECT s1_id, cand_id, COUNT(*)::INT AS shared FROM (
            SELECT s.entity_id AS s1_id, c.entity_id AS cand_id
            FROM (SELECT entity_id, country, postcode, UNNEST(name_tokens) AS tok
                  FROM s1 WHERE postcode <> '') s
            JOIN (SELECT entity_id, country, postcode, UNNEST(name_tokens) AS tok
                  FROM cand WHERE postcode <> '') c
              ON c.country = s.country AND c.postcode = s.postcode AND c.tok = s.tok
        ) GROUP BY 1, 2""")

    # ---- union, ranked by IDF CONTAINMENT (what fraction of the S1 name's
    # information content a candidate covers) rather than raw matched weight ----
    _step(con, "s1 idf norm", """
        CREATE OR REPLACE TABLE s1_norm AS
        SELECT s.entity_id, SUM(COALESCE(t.idf, 0.0)) AS total_idf
        FROM (SELECT entity_id, country, UNNEST(name_tokens) AS tok FROM s1) s
        LEFT JOIN tokdf t ON t.country = s.country AND t.tok = s.tok
        GROUP BY 1""")
    _step(con, "union (uncapped)", """
        CREATE OR REPLACE TABLE upairs AS
        SELECT u.*, u.idf_score / NULLIF(n.total_idf, 0) AS containment,
               ROW_NUMBER() OVER (PARTITION BY u.s1_id ORDER BY
                   GREATEST(u.by_a, u.by_a2) DESC,
                   u.idf_score / NULLIF(n.total_idf, 0) DESC,
                   u.shared DESC, u.idf_score DESC, u.cand_id) AS rnk
        FROM (
            SELECT s1_id, cand_id, MAX(by_a) by_a, MAX(by_a2) by_a2, MAX(by_b) by_b,
                   MAX(by_c) by_c, MAX(by_d) by_d, MAX(by_e) by_e, MAX(by_f) by_f,
                   MAX(idf_score) idf_score, MAX(shared) shared
            FROM (
                SELECT s1_id, cand_id, 1 by_a, 0 by_a2, 0 by_b, 0 by_c, 0 by_d, 0 by_e,
                       0 by_f, 0.0 idf_score, 0 shared FROM pairs_a
                UNION ALL SELECT s1_id,cand_id,0,1,0,0,0,0,0,0.0,0 FROM pairs_a2
                UNION ALL SELECT s1_id,cand_id,0,0,1,0,0,0,0,idf_score,shared FROM pairs_b
                UNION ALL SELECT s1_id,cand_id,0,0,0,1,0,0,0,idf_score,shared FROM pairs_c
                UNION ALL SELECT s1_id,cand_id,0,0,0,0,1,0,0,0.0,shared FROM pairs_d
                UNION ALL SELECT s1_id,cand_id,0,0,0,0,0,1,0,idf_score,shared FROM pairs_e
                UNION ALL SELECT s1_id,cand_id,0,0,0,0,0,0,1,idf_score,shared FROM pairs_f
            ) GROUP BY s1_id, cand_id
        ) u LEFT JOIN s1_norm n ON n.entity_id = u.s1_id""")
    _step(con, "top-k cut", f"CREATE OR REPLACE TABLE candidates AS "
                            f"SELECT * FROM upairs WHERE rnk <= {top_k}")

    out = WORK_DIR / f"{split}_candidates{suffix}.parquet"
    con.execute(f"COPY candidates TO '{out}' (FORMAT PARQUET, COMPRESSION ZSTD)")
    for t in ("pairs_a", "pairs_a2", "pairs_b", "pairs_c", "pairs_d", "pairs_e",
              "pairs_f", "upairs", "candidates"):
        print(f"  {t:<12s} {con.sql(f'SELECT COUNT(*) FROM {t}').fetchone()[0]:>13,}")
    return out

### 5.3 Run blocking on the dev sample

2% of S1 against the **whole** pool. ~75 s.

In [14]:
con = connect()
t0 = time.perf_counter()
print("--- pool index (built once) ---")
build_pool(con, "train")
print("\n--- shard: 2% of S1 ---")
cand_path = build_shard(con, "train", f"WHERE hash(entity_id) % 1000 < {CONFIG['s1_sample']}")
print(f"\ntotal {time.perf_counter() - t0:.1f}s -> {cand_path.name}")

--- pool index (built once) ---
  pool 10,320,219 records


  name tokens               1.0s
  token df                  0.7s
  squash keys               0.9s
  skeletons                 8.6s
  address tokens            5.9s
  token pairs              47.0s

--- shard: 2% of S1 ---
  S1 rows 43,955
  s1 rare tokens            0.5s
  estimate scheme B        6,116,611 pairs  [OK]
  s1 skeletons              0.4s
  estimate scheme C        5,712,701 pairs  [OK]
  s1 rare addr              0.5s
  estimate scheme E        2,568,182 pairs  [OK]
  s1 token pairs            0.4s
  estimate scheme F       10,022,418 pairs  [OK]
  A exact key               1.9s
  A2 squash                23.3s
  b rare token              7.2s
  c skeleton                2.5s
  e rare addr              15.7s
  f token pair             14.2s
  D postcode                1.3s
  s1 idf norm               0.3s
  union (uncapped)          6.8s
  top-k cut                 0.4s
  pairs_a          1,532,036
  pairs_a2         1,534,036
  pairs_b          1,784,683
  pairs_c      

### 5.4 Blocking diagnostics — the gate

`recall@K` separates the only two failure modes that matter:

- **uncapped low** -> *generation* problem: true pairs are never produced. Add a scheme.
- **uncapped high but recall@top_k low** -> *ranking* problem: they are produced and then
  discarded by the cut. Fix the ordering or raise K.

Getting this backwards wastes hours. Early on the numbers *looked* like a ranking failure
(~1,100 raw candidates per entity, cut to 30) but uncapped recall was only 0.699 — it was
generation all along, and no reordering could have reached the missing 30%.

In [15]:
def diagnose(con, ks=(10, 30, 50, 100, 200, 500)):
    gt_path = str(DATA_DIR / "train" / GROUND_TRUTH)
    con.execute(f"""
        CREATE OR REPLACE TABLE gt AS
        SELECT g.s1_id, TRIM(g.cand_id) AS cand_id FROM (
            SELECT source1_entity_id AS s1_id,
                   UNNEST(STRING_SPLIT(matched_entity_ids, ',')) AS cand_id
            FROM read_csv('{gt_path}', delim='\t', header=true,
                 columns={{'source1_entity_id':'VARCHAR','matched_entity_ids':'VARCHAR'}})
            WHERE matched_entity_ids IS NOT NULL AND matched_entity_ids <> ''
        ) g JOIN s1 ON s1.entity_id = g.s1_id""")
    cols = ", ".join(f"SUM(CASE WHEN rnk<={k} THEN 1 ELSE 0 END)*1.0/COUNT(*) AS r{k}"
                     for k in ks)
    row = con.sql(f"""
        WITH m AS (SELECT g.s1_id, g.cand_id, u.rnk FROM gt g
                   LEFT JOIN upairs u ON u.s1_id=g.s1_id AND u.cand_id=g.cand_id)
        SELECT COUNT(*),
               SUM(CASE WHEN rnk IS NOT NULL THEN 1 ELSE 0 END)*1.0/COUNT(*), {cols}
        FROM m""").fetchone()
    print(f"  true pairs            {row[0]:,}")
    print(f"  uncapped (generation) {row[1]:.4f}   <- hard ceiling of this candidate set")
    for k, v in zip(ks, row[2:]):
        print(f"  recall@{k:<4d}          {v:.4f}")
    print("  -> " + ("GENERATION is the constraint: add a scheme" if row[1] < 0.90 else
                     "RANKING/K is the constraint: found then discarded"))

def blocking_report(candidates: dict, truth: dict):
    n = len(truth); tp = cp = hits = 0
    rec = perfect = zero = ceil_sum = 0.0
    n_single = single_nocand = 0
    for sid, tset in truth.items():
        cand = candidates.get(sid, set()); hit = tset & cand
        cp += len(cand); tp += len(tset); hits += len(hit)
        if tset:
            r = len(hit) / len(tset); rec += r; perfect += (r == 1.0); zero += (r == 0.0)
        else:
            n_single += 1; single_nocand += (len(cand) == 0); rec += 1.0; perfect += 1.0
        ceil_sum += f05(hit, tset)
    print(f"  entities              {n:,}")
    print(f"  mean candidates/ent   {cp / n:.1f}")
    print(f"  recall (micro)        {hits / tp:.4f}")
    print(f"  recall (macro)        {rec / n:.4f}")
    print(f"  entities fully cover  {perfect / n:.4f}")
    print(f"  entities zero recall  {zero / max(1, n - n_single):.4f}")
    print(f"  CEILING macro F_0.5   {ceil_sum / n:.4f}   <- best any model could score here")

print("=== RECALL@K ===");  diagnose(con)
rows = con.sql("SELECT s1_id, cand_id FROM candidates").fetchnumpy()
cands = {}
for s, c in zip(rows["s1_id"], rows["cand_id"]):
    cands.setdefault(s, set()).add(c)
s1_ids = set(con.sql("SELECT entity_id FROM s1").fetchnumpy()["entity_id"])
print("\n=== BLOCKING REPORT ===");  blocking_report(cands, load_ground_truth(s1_ids))

=== RECALL@K ===
  true pairs            151,550
  uncapped (generation) 0.7982   <- hard ceiling of this candidate set
  recall@10            0.5710
  recall@30            0.6682
  recall@50            0.7034
  recall@100           0.7656
  recall@200           0.7895
  recall@500           0.7982
  -> GENERATION is the constraint: add a scheme

=== BLOCKING REPORT ===
  entities              43,955
  mean candidates/ent   28.1
  recall (micro)        0.6682
  recall (macro)        0.6874
  entities fully cover  0.4674
  entities zero recall  0.1543
  CEILING macro F_0.5   0.7926   <- best any model could score here


### 5.5 Inspect what blocking *missed*

Guessing at the cause of a recall gap is how you add the wrong scheme. Every scheme above
was chosen from this output. It is also how the address signal was found: true pairs whose
names share nothing (`bauer international` / `novimira`) still shared 4+ address tokens.

In [16]:
def inspect_misses(con, n=15):
    pool = [str(WORK_DIR / "train_s2.parquet"), str(WORK_DIR / "train_s3.parquet")]
    s1p = str(WORK_DIR / "train_s1.parquet")
    rows = con.sql(f"""
        WITH miss AS (SELECT g.s1_id, g.cand_id FROM gt g
                      LEFT JOIN upairs u ON u.s1_id=g.s1_id AND u.cand_id=g.cand_id
                      WHERE u.s1_id IS NULL)
        SELECT s.name_text, p.name_text, s.name_tokens, p.name_tokens, s.country,
               s.postcode, p.postcode, s.addr_tokens, p.addr_tokens
        FROM miss m JOIN read_parquet('{s1p}') s ON s.entity_id=m.s1_id
        JOIN read_parquet({pool}) p ON p.entity_id=m.cand_id
        USING SAMPLE {n} ROWS""").fetchall()
    shared_any = rescuable = 0
    for sn, cn, st, ct, country, sp, cp_, sat, cat in rows:
        ov = set(st or ()) & set(ct or ()); aov = set(sat or ()) & set(cat or ())
        pin = bool(sp and sp == cp_); resc = bool(pin or len(aov) >= 2)
        shared_any += bool(ov); rescuable += resc and not ov
        print(f"  [{country:6s}] {sn!r:36s} vs {cn!r:36s} "
              f"{'shared=' + str(sorted(ov)) if ov else 'NO shared name token'}")
        print(f"           addr_overlap={sorted(aov)[:4]} pin={'=' if pin else 'x'}"
              f"{'  <- ADDRESS COULD RESCUE' if resc and not ov else ''}")
    print(f"\n  {shared_any}/{len(rows)} share a name token; "
          f"{rescuable} of the rest are reachable via ADDRESS")
    print("  shares a token but missed -> a df ceiling / rarest-N / key-cap dropped it")

inspect_misses(con)

  [India ] 'shakti krishna ventures'            vs '5hakti krishna ventures'            shared=['krishna', 'ventures']
           addr_overlap=['maharashtra', 'mumbai'] pin=x
  [India ] 'great solutions'                    vs 'gret solutions'                     shared=['solutions']
           addr_overlap=['block', 'chowk', 'delhi', 'ground'] pin=x
  [India ] 'green logistics'                    vs 'grin lajistiks'                     NO shared name token
           addr_overlap=['chand', 'delhi', 'ghosh', 'gyan'] pin=x  <- ADDRESS COULD RESCUE
  [US    ] 'reix'                               vs 'verasyn'                            NO shared name token
           addr_overlap=['court', 'fairfield', 'gillotti', 'new'] pin=x  <- ADDRESS COULD RESCUE
  [US    ] 'blue pub'                           vs 'yumavio'                            NO shared name token
           addr_overlap=['elderberry', 'il', 'naperville'] pin=x  <- ADDRESS COULD RESCUE
  [US    ] 'community chiropractic'        

## 6. Stage 2 — pair features

Four families. The **group-relative** block matters more than it looks: scoring is
per-entity, so what decides a match is how a candidate compares to its siblings, not its
absolute similarity.

Column positions come from `IDX`, never hard-coded — the offsets silently drifted once
when blocking gained new provenance flags.

In [17]:
FEATURE_COLUMNS = [
    # name
    "n_token_set", "n_token_sort", "n_partial", "n_jaro", "n_prefix", "n_jacc",
    "n_idf_cover_s1", "n_idf_cover_c", "n_len_ratio", "n_tok_diff", "n_exact_key",
    "n_suffix_agree", "n_suffix_conflict",
    # address
    "a_token_set", "a_jacc", "a_post_match", "a_post_both", "a_num_overlap",
    "a_num_both", "a_empty_either",
    # blocking provenance — which scheme(s) found this pair is strong signal
    "b_by_a", "b_by_a2", "b_by_b", "b_by_c", "b_by_d", "b_by_e", "b_by_f",
    "b_n_schemes", "b_idf_score", "b_shared", "b_containment", "b_rank",
    # group-relative
    "g_n_cands", "g_rank_frac", "g_score_ratio", "g_score_margin", "g_is_top",
    # context
    "c_source", "c_same_country",
]
IDX = {c: i for i, c in enumerate(FEATURE_COLUMNS)}
N_FEATURES = len(FEATURE_COLUMNS)

def _jacc(a, b):
    return len(a & b) / len(a | b) if a and b else 0.0

def pair_features(df, idf):
    out = np.zeros((len(df), N_FEATURES), dtype=np.float32); I = IDX
    for i, r in enumerate(df.itertuples(index=False)):
        sn, cn = r.s_name_text or "", r.c_name_text or ""
        st = set(r.s_name_tokens) if r.s_name_tokens is not None else set()
        ct = set(r.c_name_tokens) if r.c_name_tokens is not None else set()
        country = r.s_country

        out[i, I["n_token_set"]] = fuzz.token_set_ratio(sn, cn) / 100.0
        out[i, I["n_token_sort"]] = fuzz.token_sort_ratio(sn, cn) / 100.0
        out[i, I["n_partial"]] = fuzz.partial_ratio(sn, cn) / 100.0
        out[i, I["n_jaro"]] = JaroWinkler.similarity(sn, cn)
        out[i, I["n_prefix"]] = fuzz.QRatio(sn[:12], cn[:12]) / 100.0
        out[i, I["n_jacc"]] = _jacc(st, ct)
        # IDF-weighted coverage both ways: rare shared tokens are far more
        # informative than common ones, which plain Jaccard cannot express.
        ws = sum(idf.get((country, t), 0.0) for t in st & ct)
        w_s = sum(idf.get((country, t), 0.0) for t in st)
        w_c = sum(idf.get((country, t), 0.0) for t in ct)
        out[i, I["n_idf_cover_s1"]] = ws / w_s if w_s else 0.0
        out[i, I["n_idf_cover_c"]] = ws / w_c if w_c else 0.0
        ls, lc = len(sn), len(cn)
        out[i, I["n_len_ratio"]] = min(ls, lc) / max(ls, lc) if max(ls, lc) else 0.0
        out[i, I["n_tok_diff"]] = abs(len(st) - len(ct))
        out[i, I["n_exact_key"]] = float(r.s_name_key == r.c_name_key and r.s_name_key != "")
        ssuf, csuf = set((r.s_name_suffix or "").split()), set((r.c_name_suffix or "").split())
        out[i, I["n_suffix_agree"]] = float(bool(ssuf & csuf))
        # Both declare a legal form and they disagree -> weak evidence of a
        # DIFFERENT company. This is why suffixes stay out of the core tokens.
        out[i, I["n_suffix_conflict"]] = float(bool(ssuf) and bool(csuf) and not (ssuf & csuf))

        sa, ca = r.s_addr_text or "", r.c_addr_text or ""
        sat = set(r.s_addr_tokens) if r.s_addr_tokens is not None else set()
        cat = set(r.c_addr_tokens) if r.c_addr_tokens is not None else set()
        out[i, I["a_token_set"]] = fuzz.token_set_ratio(sa, ca) / 100.0 if sa and ca else 0.0
        out[i, I["a_jacc"]] = _jacc(sat, cat)
        sp, cp_ = r.s_postcode or "", r.c_postcode or ""
        out[i, I["a_post_match"]] = float(sp != "" and sp == cp_)
        out[i, I["a_post_both"]] = float(sp != "" and cp_ != "")
        snum = set(r.s_addr_nums) if r.s_addr_nums is not None else set()
        cnum = set(r.c_addr_nums) if r.c_addr_nums is not None else set()
        out[i, I["a_num_overlap"]] = _jacc(snum, cnum)
        out[i, I["a_num_both"]] = float(bool(snum) and bool(cnum))
        out[i, I["a_empty_either"]] = float(r.s_addr_empty or r.c_addr_empty)

        a, a2, b, c, d, e, f = r.by_a, r.by_a2, r.by_b, r.by_c, r.by_d, r.by_e, r.by_f
        for nm, v in (("b_by_a", a), ("b_by_a2", a2), ("b_by_b", b), ("b_by_c", c),
                      ("b_by_d", d), ("b_by_e", e), ("b_by_f", f)):
            out[i, I[nm]] = v
        out[i, I["b_n_schemes"]] = a + a2 + b + c + d + e + f   # cross-scheme agreement
        out[i, I["b_idf_score"]] = r.idf_score or 0.0
        out[i, I["b_shared"]] = r.shared or 0
        out[i, I["b_containment"]] = r.containment or 0.0
        out[i, I["b_rank"]] = r.rnk
        out[i, I["c_source"]] = 2.0 if str(r.cand_id).startswith("S2-") else 3.0
        out[i, I["c_same_country"]] = float(r.s_country == r.c_country)
    return out

def add_group_features(feats, s1_ids):
    """Rank/margin within each entity's own candidate list, in place."""
    base = feats[:, IDX["n_token_set"]] * 0.6 + feats[:, IDX["n_idf_cover_s1"]] * 0.4
    g = pd.DataFrame({"sid": s1_ids, "score": base}).groupby("sid")["score"]
    n = g.transform("size").to_numpy(np.float32)
    mx = g.transform("max").to_numpy(np.float32)
    rank = g.rank(ascending=False, method="first").to_numpy(np.float32)
    feats[:, IDX["g_n_cands"]] = n
    feats[:, IDX["g_rank_frac"]] = rank / np.maximum(n, 1)
    feats[:, IDX["g_score_ratio"]] = np.where(mx > 0, base / mx, 0.0)
    feats[:, IDX["g_score_margin"]] = base - mx
    feats[:, IDX["g_is_top"]] = (rank == 1).astype(np.float32)

In [18]:
def build_features(split, suffix="", chunk=500_000):
    con = connect()
    cand = str(WORK_DIR / f"{split}_candidates{suffix}.parquet")
    s1p = str(WORK_DIR / f"{split}_s1.parquet")
    pool = [str(WORK_DIR / f"{split}_s2.parquet"), str(WORK_DIR / f"{split}_s3.parquet")]
    con.execute(f"""CREATE OR REPLACE TABLE tokdf AS
        WITH ct AS (SELECT country, UNNEST(name_tokens) AS tok FROM read_parquet({pool}))
        SELECT country, tok, LN(1.0 + (SELECT COUNT(*) FROM ct)/COUNT(*)) AS idf
        FROM ct GROUP BY country, tok""")
    t = con.sql("SELECT country, tok, idf FROM tokdf").fetchnumpy()
    idf = {(a, b): float(c) for a, b, c in zip(t["country"], t["tok"], t["idf"])}
    print(f"  idf vocabulary {len(idf):,}")

    # ORDER BY s1_id keeps each entity's candidates inside one chunk, which the
    # group-relative features require.
    con.execute(f"""CREATE OR REPLACE VIEW joined AS
        SELECT c.s1_id, c.cand_id, c.by_a, c.by_a2, c.by_b, c.by_c, c.by_d, c.by_e,
               c.by_f, c.idf_score, c.shared, c.containment, c.rnk,
               s.country s_country, s.name_key s_name_key, s.name_text s_name_text,
               s.name_tokens s_name_tokens, s.name_suffix s_name_suffix,
               s.addr_text s_addr_text, s.addr_tokens s_addr_tokens,
               s.addr_nums s_addr_nums, s.postcode s_postcode, s.addr_empty s_addr_empty,
               p.country c_country, p.name_key c_name_key, p.name_text c_name_text,
               p.name_tokens c_name_tokens, p.name_suffix c_name_suffix,
               p.addr_text c_addr_text, p.addr_tokens c_addr_tokens,
               p.addr_nums c_addr_nums, p.postcode c_postcode, p.addr_empty c_addr_empty
        FROM read_parquet('{cand}') c
        JOIN read_parquet('{s1p}') s ON s.entity_id = c.s1_id
        JOIN read_parquet({pool}) p ON p.entity_id = c.cand_id
        ORDER BY c.s1_id""")

    out = WORK_DIR / f"{split}_features{suffix}.parquet"
    schema = pa.schema([("s1_id", pa.string()), ("cand_id", pa.string())]
                       + [(c, pa.float32()) for c in FEATURE_COLUMNS])
    writer = pq.ParquetWriter(out, schema, compression="zstd")
    total, t0 = 0, time.perf_counter()
    try:
        for batch in con.sql("SELECT * FROM joined").fetch_record_batch(chunk):
            df = batch.to_pandas()
            feats = pair_features(df, idf)
            add_group_features(feats, df["s1_id"].to_numpy())
            writer.write_table(pa.table(
                {"s1_id": pa.array(df["s1_id"]), "cand_id": pa.array(df["cand_id"]),
                 **{c: pa.array(feats[:, j], pa.float32())
                    for j, c in enumerate(FEATURE_COLUMNS)}}, schema=schema))
            total += len(df)
            print(f"    {total:>12,} pairs ({total/(time.perf_counter()-t0):,.0f}/s)",
                  end="\r", flush=True)
    finally:
        writer.close()
    print(f"\n  {total:,} pairs in {time.perf_counter()-t0:.1f}s -> {out.name}")
    con.close()
    return out

build_features("train")

  ! only 13.6GB free, capping spill at 11GB
  idf vocabulary 1,483,134


/tmp/ipykernel_86609/647741998.py:38: DeprecationWarning: fetch_record_batch() is deprecated, use to_arrow_reader() instead.
  for batch in con.sql("SELECT * FROM joined").fetch_record_batch(chunk):


       1,234,626 pairs (28,984/s)
  1,234,626 pairs in 42.9s -> train_features.parquet


PosixPath('/home/subhrajit/Desktop/ML_Challenge_2026/work/train_features.parquet')

## 7. Stage 4 — set selection (defined before training, since evaluation uses it)

A single global probability threshold is the obvious approach and it is suboptimal,
because `m` varies from 0 to 11 per entity. Instead maximize **expected F_0.5** per
entity. With calibrated probabilities sorted descending:

$$E[h]_k = \sum_{i \le k} p_i, \qquad E[m] = \sum_i p_i, \qquad
\text{score}(k) = \frac{1.25 \sum_{i\le k} p_i}{k + 0.25 E[m]}$$

and $\text{score}(0) = \prod_i (1 - p_i)$, the probability the entity is genuinely a
singleton. Since $E[m]$ is constant in $k$, this is a **self-calibrating per-entity
threshold** — admit candidate $k+1$ only while its probability beats the running average.
No cutoff to tune, and it adapts to singletons and 11-match entities alike.

*Measured honestly:* this gained only **+0.004** over a tuned global threshold. When the
classifier is this confident (PR-AUC 0.99) probabilities sit near 0/1 and any sensible
threshold picks nearly the same sets. Keep it — it is free and needs no tuning — but it
is not the differentiator it looks like.

In [19]:
def select_sets(df, prob_col="p", e_m=None):
    d = df[["s1_id", "cand_id", prob_col]].sort_values(
        ["s1_id", prob_col], ascending=[True, False], kind="stable")
    p = d[prob_col].to_numpy(np.float64)
    g = d.groupby("s1_id", sort=False)
    k = g.cumcount().to_numpy() + 1
    cum_p = g[prob_col].cumsum().to_numpy(np.float64)
    # E[m] must cover EVERY candidate, including low-probability ones dropped before
    # assembly, so it can be supplied from full pre-filter sums.
    e_m_arr = (g[prob_col].transform("sum").to_numpy(np.float64) if e_m is None
               else d["s1_id"].map(e_m).fillna(0.0).to_numpy(np.float64))
    d["score"] = 1.25 * cum_p / (k + 0.25 * e_m_arr)
    d["k"] = k
    d["log1mp"] = np.log1p(-np.clip(p, 0.0, 1.0 - 1e-12))   # logs avoid underflow

    best_idx = d.groupby("s1_id", sort=False)["score"].idxmax()
    best = d.loc[best_idx, ["s1_id", "score", "k"]].set_index("s1_id")
    score0 = np.exp(d.groupby("s1_id", sort=False)["log1mp"].sum())
    kstar = best["k"].where(~(score0 > best["score"]), 0)

    d["kstar"] = d["s1_id"].map(kstar).to_numpy()
    sel = d[d["k"] <= d["kstar"]]
    out = {sid: set() for sid in kstar.index}
    for sid, cid in zip(sel["s1_id"].to_numpy(), sel["cand_id"].to_numpy()):
        out[sid].add(cid)
    return out

def select_global_threshold(df, thr, prob_col="p"):
    """Baseline for comparison: one cutoff for every entity."""
    out = {sid: set() for sid in df["s1_id"].unique()}
    hit = df[df[prob_col] >= thr]
    for sid, cid in zip(hit["s1_id"].to_numpy(), hit["cand_id"].to_numpy()):
        out[sid].add(cid)
    return out

## 8. Stage 5 — global exclusivity repair

§2.1 verified that each S2/S3 record serves at most one S1 entity, so a contested id
means at least one claim is wrong — and under a precision-heavy metric dropping the
weaker claim is nearly free. Removing a candidate changes that entity's optimal `k`, so
selection is re-run on the survivors and the process iterates.

This must run **once, globally**: if two entities in different shards both claim `S2-x`,
per-shard repair would never see the conflict.

In [20]:
def repair_exclusivity(pred, df, prob_col="p", rounds=3, e_m=None):
    prob = {(s, c): p for s, c, p in zip(df["s1_id"].to_numpy(),
                                        df["cand_id"].to_numpy(), df[prob_col].to_numpy())}
    total = 0
    for _ in range(rounds):
        owners = {}
        for sid, cands in pred.items():
            for cid in cands:
                owners.setdefault(cid, []).append(sid)
        contested = {c: s for c, s in owners.items() if len(s) > 1}
        if not contested:
            break
        total += len(contested)
        drop = set()
        for cid, sids in contested.items():
            winner = max(sids, key=lambda s: prob.get((s, cid), 0.0))
            drop.update((s, cid) for s in sids if s != winner)
        surviving = df[~pd.MultiIndex.from_arrays([df["s1_id"], df["cand_id"]]).isin(drop)]
        pred = select_sets(surviving, prob_col, e_m)
    return pred, total

## 9. Stage 3 — train the pair scorer, then evaluate end to end

Folds are split **by S1 entity**, three ways (70 train / 15 calib / 15 val), so
calibration and scoring never share a fold. The candidate pool is never shrunk, so the
val score is capped by real blocking recall rather than an easy subset — the most common
way to build a validation number that is 5 points optimistic.

Negatives are never subsampled per entity: ~26% of the pool matches nothing, and that
distractor mix is what teaches precision.

In [21]:
MODEL_PATH, CALIB_PATH = WORK_DIR / "lgbm.txt", WORK_DIR / "calib.npz"

def fold_of(s1_ids):
    h = pd.util.hash_array(np.asarray(s1_ids), hash_key="0" * 16) % 100
    return np.where(h < 70, 0, np.where(h < 85, 1, 2))

def train_model(split="train"):
    import lightgbm as lgb
    from sklearn.isotonic import IsotonicRegression
    from sklearn.metrics import average_precision_score, roc_auc_score

    feats = pd.read_parquet(WORK_DIR / f"{split}_features.parquet")
    truth_all = load_ground_truth(set(feats["s1_id"].unique()))
    pos = {(s, c) for s, cs in truth_all.items() for c in cs}
    feats["y"] = [int((s, c) in pos) for s, c in
                  zip(feats["s1_id"].to_numpy(), feats["cand_id"].to_numpy())]
    feats["fold"] = fold_of(feats["s1_id"].to_numpy())
    print(f"  pairs {len(feats):,}  positives {feats.y.sum():,} ({100*feats.y.mean():.2f}%)")

    tr, ca, va = feats[feats.fold == 0], feats[feats.fold == 1], feats[feats.fold == 2]
    print(f"  entities  train {tr.s1_id.nunique():,}  calib {ca.s1_id.nunique():,}"
          f"  val {va.s1_id.nunique():,}")

    dtr = lgb.Dataset(tr[FEATURE_COLUMNS], tr["y"])
    dca = lgb.Dataset(ca[FEATURE_COLUMNS], ca["y"], reference=dtr)
    booster = lgb.train(
        dict(objective="binary", metric="average_precision", learning_rate=0.05,
             num_leaves=63, min_data_in_leaf=100, feature_fraction=0.9,
             bagging_fraction=0.8, bagging_freq=1, seed=42, num_threads=CONFIG["threads"],
             verbosity=-1, deterministic=True),  # seed+deterministic: the
             # submission must be reproducible, and bagging without a
             # seed drifted the val score by ~0.002 between runs
        dtr, num_boost_round=600, valid_sets=[dca],
        callbacks=[lgb.early_stopping(40, verbose=False), lgb.log_evaluation(200)])
    print(f"  best iteration {booster.best_iteration}")

    raw_ca = booster.predict(ca[FEATURE_COLUMNS], num_iteration=booster.best_iteration)
    iso = IsotonicRegression(out_of_bounds="clip").fit(raw_ca, ca["y"])
    raw_va = booster.predict(va[FEATURE_COLUMNS], num_iteration=booster.best_iteration)
    va = va.copy(); va["p"] = iso.predict(raw_va)
    print(f"  val ROC-AUC {roc_auc_score(va.y, raw_va):.4f}  "
          f"PR-AUC {average_precision_score(va.y, raw_va):.4f}")

    booster.save_model(str(MODEL_PATH), num_iteration=booster.best_iteration)
    np.savez(CALIB_PATH, x=iso.X_thresholds_, y=iso.y_thresholds_)
    return booster, iso, va

booster, iso, va = train_model()

  pairs 1,234,626  positives 101,262 (8.20%)
  entities  train 30,627  calib 6,642  val 6,677
[200]	valid_0's average_precision: 0.989726
[400]	valid_0's average_precision: 0.990058
  best iteration 416
  val ROC-AUC 0.9991  PR-AUC 0.9901


In [22]:
truth = load_ground_truth(set(va["s1_id"].unique()))

best_thr, best_score = 0.5, -1.0
for thr in np.arange(0.05, 0.96, 0.05):
    s = macro_f05(select_global_threshold(va, thr), truth)
    if s > best_score:
        best_thr, best_score = thr, s

sel = select_sets(va)
exp_score = macro_f05(sel, truth)
repaired, conflicts = repair_exclusivity(sel, va)
rep_score = macro_f05(repaired, truth)

print("=== VAL MACRO F_0.5 ===")
print(f"  global threshold (best {best_thr:.2f})   {best_score:.4f}")
print(f"  expected-F_0.5 selection          {exp_score:.4f}  ({exp_score-best_score:+.4f})")
print(f"  + exclusivity repair              {rep_score:.4f}  ({rep_score-exp_score:+.4f}, "
      f"{conflicts:,} contested)")

imp = pd.Series(booster.feature_importance("gain"),
                index=FEATURE_COLUMNS).sort_values(ascending=False)
print("\n=== TOP FEATURES (gain) ===")
for k, v in imp.head(12).items():
    print(f"  {k:<18s} {v:12,.0f}")

=== VAL MACRO F_0.5 ===
  global threshold (best 0.65)   0.7392
  expected-F_0.5 selection          0.7410  (+0.0018)
  + exclusivity repair              0.7410  (+0.0000, 0 contested)

=== TOP FEATURES (gain) ===
  a_token_set           1,884,476
  a_num_overlap           646,698
  a_jacc                  545,478
  n_partial               134,288
  n_idf_cover_c            78,526
  n_suffix_conflict        67,555
  a_num_both               60,553
  b_rank                   58,897
  n_jaro                   33,247
  b_idf_score              31,915
  n_len_ratio              30,092
  g_rank_frac              28,145


## 10. Test inference + submission

The S1 side is **sharded**: the union's window function scales with S1 size and OOMs well
before 1.73M entities fit in one pass. The pool index is built once and reused.

Per shard: block -> features -> score, keeping only pairs with `p >= 0.01` (they can
never be selected) plus the **exact pre-filter per-entity sum** for `E[m]`. Feature files
are deleted as we go — at ~30 candidates/entity the full set is several GB.

Stages 4 and 5 then run **once, over all shards together**.

This is the slow cell (~1 h for 8 shards). Run it when you want a submission.

In [ ]:
PROB_FLOOR = 0.01

def predict_shard(split, suffix, booster, iso):
    feats = pd.read_parquet(WORK_DIR / f"{split}_features{suffix}.parquet")
    p = iso.predict(booster.predict(feats[FEATURE_COLUMNS]))
    out = pd.DataFrame({"s1_id": feats.s1_id, "cand_id": feats.cand_id, "p": p})
    sums = out.groupby("s1_id", sort=False)["p"].sum().rename("e_m").reset_index()
    kept = out[out.p >= PROB_FLOOR]
    kept.to_parquet(WORK_DIR / f"{split}_probs{suffix}.parquet", index=False)
    sums.to_parquet(WORK_DIR / f"{split}_sums{suffix}.parquet", index=False)
    print(f"  {suffix}: scored {len(out):,} -> kept {len(kept):,}, {len(sums):,} entities")

def run_test(shards=None, resume=True):
    """Sharded test inference. Resumable: a shard whose probs+sums+candidates all
    exist is skipped, so a crash costs one shard, not the whole hour."""
    shards = shards or CONFIG["test_shards"]
    if not resume:
        for pat in ("test_probs*", "test_sums*", "test_candidates*"):
            for f in WORK_DIR.glob(f"{pat}.parquet"):
                f.unlink()
    for f in WORK_DIR.glob("test_features*.parquet"):
        f.unlink()          # always stale; rebuilt per shard

    def done(sfx):
        return all((WORK_DIR / f"test_{k}{sfx}.parquet").exists()
                   for k in ("probs", "sums", "candidates"))

    todo = [i for i in range(shards) if not done(f"_sh{i:02d}")]
    if not todo:
        print("all shards already complete - go straight to assemble()")
        return
    print(f"shards to run: {todo}  (skipping {shards - len(todo)} already done)")

    dbfile = WORK_DIR / "er_test.duckdb"
    for suf in ("", ".wal"):
        (WORK_DIR / f"er_test.duckdb{suf}").unlink(missing_ok=True)
    con = connect(db=dbfile)
    print("--- pool index (once, on disk) ---"); build_pool(con, "test")
    for i in todo:
        sfx = f"_sh{i:02d}"
        print(f"\n--- shard {i+1}/{shards} ---")
        build_shard(con, "test", f"WHERE hash(entity_id) % {shards} = {i}", sfx)
        # Free the shard's tables before the next one; the pool index stays.
        for t in ("upairs", "candidates", "s1_rare", "s1_skel", "s1_arare", "s1_rare4",
                  "s1_pair", "s1_norm", "pairs_a", "pairs_a2", "pairs_b", "pairs_c",
                  "pairs_d", "pairs_e", "pairs_f"):
            con.execute(f"DROP TABLE IF EXISTS {t}")
        con.execute("CHECKPOINT")
        build_features("test", sfx)
        predict_shard("test", sfx, booster, iso)
        (WORK_DIR / f"test_features{sfx}.parquet").unlink(missing_ok=True)
        print(f"  free disk {shutil.disk_usage(WORK_DIR).free/2**30:.1f} GB")
    con.close()

def assemble(split="test"):
    probs = pd.concat([pd.read_parquet(f) for f in
                       sorted(WORK_DIR.glob(f"{split}_probs*.parquet"))], ignore_index=True)
    sums = pd.concat([pd.read_parquet(f) for f in
                      sorted(WORK_DIR.glob(f"{split}_sums*.parquet"))],
                     ignore_index=True).groupby("s1_id", sort=False)["e_m"].sum()
    print(f"  probs {len(probs):,}  entities with candidates {len(sums):,}")

    pred = select_sets(probs, "p", sums)
    pred, conflicts = repair_exclusivity(pred, probs, "p", 3, sums)
    print(f"  contested ids resolved: {conflicts:,}")

    # EVERY S1 entity must appear exactly once, singletons included.
    all_s1 = pq.read_table(WORK_DIR / f"{split}_s1.parquet",
                           columns=["entity_id"])["entity_id"].to_pylist()
    with open(OUTPUT_DIR / "matching_results.tsv", "w") as fh:
        fh.write("source1_entity_id\tmatched_entity_ids\n")
        for sid in all_s1:
            fh.write(f"{sid}\t{','.join(sorted(pred.get(sid, ())))}\n")

    cands = {}
    for f in sorted(WORK_DIR.glob(f"{split}_candidates*.parquet")):
        t = pq.read_table(f, columns=["s1_id", "cand_id"])
        for a, b in zip(t["s1_id"].to_pylist(), t["cand_id"].to_pylist()):
            cands.setdefault(a, set()).add(b)
    with open(OUTPUT_DIR / "candidate_pairs.tsv", "w") as fh:
        fh.write("source1_entity_id\tcandidate_entity_ids\n")
        for sid in all_s1:
            fh.write(f"{sid}\t{','.join(sorted(cands.get(sid, ())))}\n")

    n = sum(1 for s in all_s1 if pred.get(s))
    print(f"  wrote both TSVs: {len(all_s1):,} rows, {n:,} with >=1 match "
          f"({100*n/len(all_s1):.1f}%)")

run_test()      # <-- uncomment: the ~1 hour full inference pass
assemble()      # <-- then this

  ! only 13.6GB free, capping spill at 11GB
--- pool index (once) ---
  pool 9,969,589 records
  name tokens               1.0s
  token df                  0.6s
  squash keys               1.0s
  skeletons                10.4s


## 11. Validate before uploading

The official `utils/validate_submission.py` checks every format rule, so a rejection
costs nothing instead of a submission. It must print **PASS**.

In [24]:
validator = DATA_DIR.parent / "utils" / "validate_submission.py"
cmd = [sys.executable, str(validator),
       "--matching", str(OUTPUT_DIR / "matching_results.tsv"),
       "--candidate", str(OUTPUT_DIR / "candidate_pairs.tsv"),
       "--test-dir", str(DATA_DIR / "test")]
print(" ".join(cmd), "\n")
r = subprocess.run(cmd, capture_output=True, text=True)
print(r.stdout[-3000:] or r.stderr[-2000:])
print(f"exit code {r.returncode}  ->  {'PASS' if r.returncode == 0 else 'FAIL'}")

/home/subhrajit/Desktop/ML_Challenge_2026/.venv/bin/python /home/subhrajit/Desktop/ML_Challenge_2026/6ab10eb3b23ba_student_resource/student_resource/utils/validate_submission.py --matching /home/subhrajit/Desktop/ML_Challenge_2026/ML_Challenge_2026/output/matching_results.tsv --candidate /home/subhrajit/Desktop/ML_Challenge_2026/ML_Challenge_2026/output/candidate_pairs.tsv --test-dir /home/subhrajit/Desktop/ML_Challenge_2026/6ab10eb3b23ba_student_resource/student_resource/dataset/test 

ML Challenge 2026 — submission validator
  test dir: /home/subhrajit/Desktop/ML_Challenge_2026/6ab10eb3b23ba_student_resource/student_resource/dataset/test
  required S1 entities: 1732544

FAIL — 1 issue(s) to fix before submitting:
  1. File not found: /home/subhrajit/Desktop/ML_Challenge_2026/ML_Challenge_2026/output/matching_results.tsv

exit code 1  ->  FAIL


## 12. Where the score stands, and what is left

Measured on a 2% S1 sample against the full pool, blocking improved across iteration 1:

| Change | generation | recall@30 | ceiling F_0.5 |
|---|---|---|---|
| baseline (A, B, D) | 0.699 | 0.558 | 0.730 |
| + A2 squash, + C skeleton, A cap fix | 0.766 | 0.588 | 0.749 |
| + per-scheme caps (120s -> 41s) | 0.698 | 0.585 | 0.749 |
| + E rare address tokens | 0.735 | 0.611 | 0.765 |
| + F conjunctive token pairs | 0.798 | 0.652 | 0.785 |
| df ceilings 5000 -> 1000 (8x cheaper) | 0.798 | **0.668** | **0.793** |

End to end: **val macro F_0.5 = 0.7433**, model PR-AUC 0.9901.

**The binding constraint is blocking recall, not the model.** 0.7433 against a 0.793
ceiling is ~94% of available headroom, and the model already separates pairs almost
perfectly. Next levers, in order:

1. **Raise `top_k` 30 -> 100.** Measured: `recall@100 = 0.766` vs `recall@30 = 0.668`.
   Costs only Stage 2 compute. Requires raising `scheme_cap` too.
2. **More address-led blocking and features.** `a_token_set` has 3.3x the gain of the next
   feature and 4 of the top 10 are address-derived.
3. Char n-gram blocking for the residue that no token or skeleton key reaches.

**The biggest unknown is France.** Training data is US + India only; the test set is 15%
French (259,452 entities) with zero training examples. Nothing here is hardcoded to
`{US, India}` — every join is generic over `country`, and the normalizer handles French
accents and `SARL`/`SAS` — but "should generalize" is not "measured to generalize". The
first leaderboard submission is the only instrument that sees France at all.